adam-core/src/main/scala/org/bdgenomics/adam/rdd/ShuffleRegionJoin.scala

```diff
@@ -38,488 +35,421 @@ import scala.reflect.ClassTag
 sealed abstract class ShuffleRegionJoin[T: ClassTag, U: ClassTag, RT, RU]
     extends RegionJoin[T, U, RT, RU] {
 
-  protected val sd: SequenceDictionary
-  protected val partitionSize: Long
-  protected val sc: SparkContext
+  protected def advanceCache(setTheoryCache: SetTheoryCache[U, RT, RU],
```

Comment:
On the note about moving the `advanceCache` and `pruneCache` into the `SetTheoryCache` class, I have given it some thought and here is how it would look.
- `advanceCache` would have to accept a function `(ReferenceRegion, ReferenceRegion) => Boolean` as a parameter, in addition to their current requirements
- `pruneCache` would have to accept 3 functions `(ReferenceRegion, ReferenceRegion) => Boolean` as parameters, in addition to their current requirements
>>This is due to the varying conditions under which you prune/advance the cache(s).

I am happy to add it, but I don't know how it will affect the performance. The complexity of the function objects being passed around should be pretty low. Let me know your thoughts on this.

 would have to accept a method 

I had it as `cache`, but it looked weird to keep using `cache.cache`. I'll put it back as `cache`, I don't have a preference either way.